README
Encode-JISX0213 package.

Copyright (C) 2013 by Hatuka*nezumi - IKEDA Soji

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

                                   ***

Two main modules and some supporting modules are contained.  For more
details read following POD documentations:

    Encode::JISX0213 - JIS X 0213 encodings
    Encode::ShiftJIS2004 - shift_jis-2004 - JIS X 0213 Annex 1 encoding

$$


Changes
Revision history for Perl extension Encode-JISX0213.

0.03  Web Sep 25 2013

	* t/JISX0213.t: RT tests had always succeeded.
	* ShiftJIS2004.pm: Deprecated PERLQQ feature of econde(),
	  and added multibyte-aware error handling to decode().
	* ucm/jis-x-0213-annex1.ucm: added fallback mappings for alternative
	  names (per Annex 5) of two characters: reverse solidus and tilde.

0.02  Mon Sep 23 2013

	* First stable (hopefully) release.

0.01_05  Sun Sep 22 2013

	* Supported enc2xs.bat for MSWin32
	* Fixed broken encoding by shift_jis-2004: Now it is provied by
	  separate package Encode::ShiftJIS2004.

0.01_02  Wed Sep 18 2013

	* jis-x-0208.ucm: removed fallback maps for CP932, except U+2015 which
	  is defined by jis0208-raw encoding of Encode core module.
	* JISX0213.pm:
	  - Introduced shortest encoding strategy for secondary CCS of
	    iso-2022-* encodings.
	  - Canonical name of "iso-2022-jp-2004-compatible" was renamed to
	    "x-iso-2022-jp-2004-compatible".
	  - Added "x-iso-2022-jp-2004-strict" encoding.

0.01_00  Tue Sep 17 2013

	* First beta.

0.000_07  Sun Sep 15 2013

	* Second CPAN upload.

0.000_06  Sun Sep 15 2013

	* First CPAN upload.



lib/Encode/JISX0213.pm
#-*- perl -*-
#-*- coding: utf-8 -*-

package Encode::JISX0213;

use strict;
use warnings;
use base qw(Encode::ISO2022);
our $VERSION = '0.03';

use Encode::ISOIRSingle;
use Encode::JISLegacy;
use Encode::JISX0213::CCS;

__PACKAGE__->Define(
    Alias => qr/\beuc-?(jis|jp)-?2004$/i,
    Name  => 'euc-jis-2004',
    CCS   => [
	{   cl       => 1,
	    encoding => 'ascii',
	    g_init   => 'g0',
	},
	{encoding => 'c1-ctrl',},
	{   bytes    => 2,
	    encoding => 'jis-x-0213-plane1-ascii',
	    gr       => 1,
	    g_init   => 'g1',
	    range    => '\xA1-\xFE',
	},
	{   bytes    => 2,
	    encoding => 'jis-x-0213-plane2',
	    gr       => 1,
	    g_init   => 'g3',
	    ss       => "\x8F",
	    range    => '\xA1-\xFE',
	},

	# Unrecommended encodings
	{   encoding => 'jis-x-0201-right',
	    gr       => 1,
	    g_init   => 'g2',
	    ss       => "\x8E",
	},

	# Nonstandard
	{   bytes    => 2,
	    encoding => 'jis-x-0212-ascii',
	    gr       => 1,
	    g_init   => 'g3',
	    ss       => "\x8F",
	    range    => '\xA1-\xFE',
	},
    ],
    SubChar => "\x{3013}",
);

Encode::define_alias(qr/\beucjisx0213$/i              => '"euc-jisx0213"');
Encode::define_alias(qr/\beuc.*jp[ \-]?(?:2000|2k)$/i => '"euc-jisx0213"');
Encode::define_alias(qr/\bjp.*euc[ \-]?(2000|2k)$/i   => '"euc-jisx0213"');
Encode::define_alias(qr/\bujis[ \-]?(?:2000|2k)$/i    => '"euc-jisx0213"');
__PACKAGE__->Define(
    Name  => 'euc-jisx0213',
    CCS   => [
	{   cl       => 1,
	    encoding => 'ascii',
	    g_init   => 'g0',
	},
	{encoding => 'c1-ctrl',},
	{   bytes    => 2,
	    encoding => 'jis-x-0213-plane1-2000-ascii',
	    gr       => 1,
	    g_init   => 'g1',
	    range    => '\xA1-\xFE',
	},
	{   bytes    => 2,
	    encoding => 'jis-x-0213-plane2',
	    gr       => 1,
	    g_init   => 'g3',
	    ss       => "\x8F",
	    range    => '\xA1-\xFE',
	},

	# Unrecommended encodings
	{   encoding => 'jis-x-0201-right',
	    gr       => 1,
	    g_init   => 'g2',
	    ss       => "\x8E",
	},

	# Nonstandard
	{   bytes    => 2,
	    encoding => 'jis-x-0212-ascii',
	    gr       => 1,
	    g_init   => 'g3',
	    ss       => "\x8F",
	    range    => '\xA1-\xFE',
	},
    ],
    SubChar => "\x{3013}",
);

__PACKAGE__->Define(
    Alias => qr/\biso-?2022-?jp-?2004$/i,
    Name  => 'iso-2022-jp-2004',
    CCS   => [
	{   cl       => 1,
	    encoding => 'ascii',
	    g_init   => 'g0',
	    g_seq    => "\e\x28\x42",
	},
	{   bytes    => 2,
	    encoding => 'jis-x-0213-plane1-ascii',
	    g        => 'g0',
	    g_seq    => "\e\x24\x28\x51",
	    range    => '\x21-\x7E',
	},
	{   bytes    => 2,
	    encoding => 'jis-x-0213-plane2',
	    g        => 'g0',
	    g_seq    => "\e\x24\x28\x50",
	    range    => '\x21-\x7E',
	},

	# Unrecommended encodings.
	{   bytes    => 2,
	    dec_only => 1,
	    encoding => 'jis-x-0213-plane1-2000-ascii',
	    g        => 'g0',
	    g_seq    => "\e\x24\x28\x4F",
	    range    => '\x21-\x7E',
	},
	{   bytes    => 2,
	    encoding => 'jis-x-0208-ascii',
	    g        => 'g0',
	    g_seq    => "\e\x24\x42",
	    ss       => '', # encodes runs as short as possible
	    range    => '\x21-\x7E',
	},

	# Nonstandard
    ],
    SubChar => "\x{3013}",
);

__PACKAGE__->Define(
    Alias => qr/\biso-?2022-?jp-?2004-?strict$/i,
    Name  => 'x-iso2022jp2004-strict',
    CCS   => [
	{   cl       => 1,
	    encoding => 'ascii',
	    g_init   => 'g0',
	    g_seq    => "\e\x28\x42",
	},
	{   bytes    => 2,
	    encoding => 'jis-x-0208-0213-ascii',
	    g        => 'g0',
	    g_seq    => "\e\x24\x42",
	    range    => '\x21-\x7E',
	},
	{   bytes    => 2,
	    encoding => 'jis-x-0213-plane1-ascii',
	    g        => 'g0',
	    g_seq    => "\e\x24\x28\x51",
	    ss       => '', # encodes runs as short as possible
	    range    => '\x21-\x7E',
	},
	{   bytes    => 2,
	    encoding => 'jis-x-0213-plane2',
	    g        => 'g0',
	    g_seq    => "\e\x24\x28\x50",
	    range    => '\x21-\x7E',
	},

	# Unrecommended encodings.
	{   bytes    => 2,
	    dec_only => 1,
	    encoding => 'jis-x-0213-plane1-2000-ascii',
	    g        => 'g0',
	    g_seq    => "\e\x24\x28\x4F",
	    range    => '\x21-\x7E',
	},

	# Nonstandard
    ],
    SubChar => "\x{3013}",
);

__PACKAGE__->Define(
    Alias => qr/\biso-?2022-?jp-?2004-?compatible$/i,
    Name  => 'x-iso2022jp2004-compatible',
    CCS   => [
	{   cl       => 1,
	    encoding => 'ascii',
	    g_init   => 'g0',
	    g_seq    => "\e\x28\x42",
	},
	{   bytes    => 2,
	    encoding => 'jis-x-0208-ascii',
	    g        => 'g0',
	    g_seq    => "\e\x24\x42",
	    range    => '\x21-\x7E',
	},
	{   bytes    => 2,
	    encoding => 'jis-x-0213-plane1-ascii',
	    g        => 'g0',
	    g_seq    => "\e\x24\x28\x51",
	    ss       => '', # encodes runs as short as possible
	    range    => '\x21-\x7E',
	},
	{   bytes    => 2,
	    encoding => 'jis-x-0213-plane2',
	    g        => 'g0',
	    g_seq    => "\e\x24\x28\x50",
	    range    => '\x21-\x7E',
	},

	# Unrecommended encodings.
	{   bytes    => 2,
	    dec_only => 1,
	    encoding => 'jis-x-0213-plane1-2000-ascii',
	    g        => 'g0',
	    g_seq    => "\e\x24\x28\x4F",
	    range    => '\x21-\x7E',
	},

	# Nonstandard
    ],
    SubChar => "\x{3013}",
);

__PACKAGE__->Define(
    Alias => qr/\biso-?2022-?jp-?3$/i,
    Name  => 'iso-2022-jp-3',
    CCS   => [
	{   cl       => 1,
	    encoding => 'ascii',
	    g_init   => 'g0',
	    g_seq    => "\e\x28\x42",
	},
	{   bytes    => 2,
	    encoding => 'jis-x-0213-plane1-2000-ascii',
	    g        => 'g0',
	    g_seq    => "\e\x24\x28\x4F",
	    range    => '\x21-\x7E',
	},
	{   bytes    => 2,
	    encoding => 'jis-x-0213-plane2',
	    g        => 'g0',
	    g_seq    => "\e\x24\x28\x50",
	    range    => '\x21-\x7E',
	},

	# Unrecommended encoding.
	{   bytes    => 2,
	    encoding => 'jis-x-0208-ascii',
	    g        => 'g0',
	    g_seq    => "\e\x24\x42",
	    ss       => '', # encodes runs as short as possible
	    range    => '\x21-\x7E',
	},

	# Nonstandard
    ],
    SubChar => "\x{3013}",
);

sub needs_lines { 1 }

1;
__END__

=encoding utf-8

=head1 NAME

Encode::JISX0213 - JIS X 0213 encodings

=head1 SYNOPSIS

    use Encode::JISX0213;
    use Encode qw/encode decode/;
    $byte = encode("iso-2022-jp-2004", $utf8);
    $utf8 = decode("iso-2022-jp-2004", $byte);

=head1 ABSTRACT

This module provides following encodings.

  Canonical         Alias                         Description
  --------------------------------------------------------------
  euc-jis-2004      qr/\beuc-?(jis|jp)-?2004$/i   8-bit encoding
  iso-2022-jp-2004  qr/\biso-?2022-?jp-?2004$/i   7-bit encoding
  --------------------------------------------------------------

=over

=item *

About "shift encoding" see L<Encode::ShiftJIS2004>.

=back

Additionally, it provides encodings for older revision, JIS X 0213:2000:

  Canonical         Alias                         Description
  --------------------------------------------------------------
  euc-jisx0213      qr/\beucjisx0213$/i           8-bit encoding
                    qr/\beuc.*jp[ \-]?(?:2000|2k)$/i
                    qr/\bjp.*euc[ \-]?(2000|2k)$/i
                    qr/\bujis[ \-]?(?:2000|2k)$/i
  iso-2022-jp-3     qr/\biso-?2022-?jp-?3$/i      7-bit encoding
  --------------------------------------------------------------

and for transition from legacy standards:

  Canonical         Alias                         Description
  --------------------------------------------------------------
  x-iso2022jp2004-compatible                      See note.
                    qr/\biso-?2022-?jp-?2004-?compatible$/i
  x-iso2022jp2004-strict                          See note.
                    qr/\biso-?2022-?jp-?2004-?strict$/i
  --------------------------------------------------------------

=head1 DESCRIPTION

To find out how to use this module in detail, see L<Encode>.

=head2 Notes on Compatibility

To encode Unicode string to byte string,
C<x-iso2022jp2004-strict> uses JIS X 0208 as much as possible,
strictly conforming to JIS X 0213:2004 Annex 2.
It is compatible to other encodings.

C<x-iso2022jp2004-compatible> uses JIS X 0208 for the bit combinations
co-existing on JIS X 0208 and JIS X 0213 plane 1.
It is I<not> compatible to other encodings;
it had never been registered by any standards bodies.

However, to decode byte string to Unicode string,
encodings in the tables above
accept arbitrary sequences of both JIS X 0208 and JIS X 0213.
Exception is C<x-iso2022jp2004-strict>:
It accepts only allowed JIS X 0208 sequences.

C<euc-jis-2004> and C<euc-jisx0213> contains JIS X 0213 plane 2 along with
JIS X 0212 in G3.
By this non-standard extension, they have forward compatibility with
EUC-JP (AJEC) in case of decoding.

=head2 Comparison with Other Modules

L<Encode::JIS2K> provides C<euc-jisx0213>, C<iso-2022-jp-3> and
C<shift_jisx0213> encodings.
C<euc-jp> in L<Encode::JP> 1.64 or later can decode C<euc-jisx0213>
sequences.
They support earlier revision of standard, and lack function to encode
sequences with multiple Unicode characters such as
katakana with semi-voiced sound mark.

L<ShiftJIS::X0213::MapUTF> provides C<shift_jisx0213> and C<shift_jis-2004>.
It does not provide interface to contemporary L<Encode> module.

=head1 SEE ALSO

JIS X 0213:2000
I<7ビット及び8ビットの2バイト情報交換用符号化拡張漢字集合>
(I<7-bit and 8-bit double byte coded extended KANJI sets for information
interchange>),
and its amendment JIS X 0213:2000/Amd.1:2004.

L<Encode>, L<Encode::JP>, L<Encode::ShiftJIS2004>.

=head1 AUTHOR

Hatuka*nezumi - IKEDA Soji <hatuka(at)nezumi.nu>

=head1 COPYRIGHT

Copyright (C) 2013 Hatuka*nezumi - IKEDA Soji.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut


lib/Encode/JISX0213/CCS.pm
#-*- perl -*-
#-*- coding: us-ascii -*-

package Encode::JISX0213::CCS;

use strict;
use warnings;
use base qw(Encode::Encoding);
our $VERSION = '0.03';

use Carp qw(carp croak);
use XSLoader;
XSLoader::load('Encode::JISX0213', $VERSION);

my $err_encode_nomap = '"\x{%*v04X}" does not map to %s';

my $DIE_ON_ERR = Encode::DIE_ON_ERR();
my $FB_QUIET = Encode::FB_QUIET();
my $HTMLCREF = Encode::HTMLCREF();
my $LEAVE_SRC = Encode::LEAVE_SRC();
my $PERLQQ = Encode::PERLQQ();
my $RETURN_ON_ERR = Encode::RETURN_ON_ERR();
my $WARN_ON_ERR = Encode::WARN_ON_ERR();
my $XMLCREF = Encode::XMLCREF();

# Workaround for encengine.c which cannot correctly map Unicode sequence
# with multiple characters.
my %composed = (
    "\x{304B}\x{309A}" => "\x24\x77",
    "\x{304D}\x{309A}" => "\x24\x78",
    "\x{304F}\x{309A}" => "\x24\x79",
    "\x{3051}\x{309A}" => "\x24\x7A",
    "\x{3053}\x{309A}" => "\x24\x7B",
    "\x{30AB}\x{309A}" => "\x25\x77",
    "\x{30AD}\x{309A}" => "\x25\x78",
    "\x{30AF}\x{309A}" => "\x25\x79",
    "\x{30B1}\x{309A}" => "\x25\x7A",
    "\x{30B3}\x{309A}" => "\x25\x7B",
    "\x{30BB}\x{309A}" => "\x25\x7C",
    "\x{30C4}\x{309A}" => "\x25\x7D",
    "\x{30C8}\x{309A}" => "\x25\x7E",
    "\x{31F7}\x{309A}" => "\x26\x78",
    "\x{00E6}\x{0300}" => "\x2B\x44",
    "\x{0254}\x{0300}" => "\x2B\x48",
    "\x{0254}\x{0301}" => "\x2B\x49",
    "\x{028C}\x{0300}" => "\x2B\x4A",
    "\x{028C}\x{0301}" => "\x2B\x4B",
    "\x{0259}\x{0300}" => "\x2B\x4C",
    "\x{0259}\x{0301}" => "\x2B\x4D",
    "\x{025A}\x{0300}" => "\x2B\x4E",
    "\x{025A}\x{0301}" => "\x2B\x4F",
    "\x{0301}"         => "\x2B\x5A",
    "\x{0300}"         => "\x2B\x5C",
    "\x{02E5}"         => "\x2B\x60",
    "\x{02E9}"         => "\x2B\x64",
    "\x{02E9}\x{02E5}" => "\x2B\x65",
    "\x{02E5}\x{02E9}" => "\x2B\x66",
);
my $composed         = join '|', reverse sort keys %composed;
my $composed_legacy  = '[^\x00-\x1F\x7F-\x9F][\x{0300}-\x{036F}\x{309A}]+';
my $prohibited_ascii = '[\x21-\x7E]';
my $prohibited_jis   = '[\x21-\x5B\x{00A5}\x5D-\x7D\x{203E}]';

foreach my $encoding (
    qw/jis-x-0208 jis-x-0208-0213 jis-x-0213-plane1 jis-x-0213-plane1-2000/
) {
    foreach my $alt ('', 'ascii', 'jis') {
	my $name = $encoding . ($alt ? "-$alt" : "");
	my $jisx0213 = ($name =~ /jis-x-0213/) ? 1 : 0;

	my $regexp;
	unless ($jisx0213) {
	    if ($alt eq 'ascii') {
		$regexp = qr{
		    \A (.*?) ($composed_legacy | $prohibited_ascii | \z)
		}osx;
	    } elsif ($alt eq 'jis') {
		$regexp = qr{
		    \A (.*?) ($composed_legacy | $prohibited_jis | \z)
		}osx;
	    } else {
		$regexp = qr{
		    \A (.*?) ($composed_legacy | \z)
		}osx;
	    }
	} else {
	    if ($alt eq 'ascii') {
		$regexp = qr{\A (.*?) ($composed | $prohibited_ascii | \z)}osx;
	    } elsif ($alt eq 'jis') {
		$regexp = qr{\A (.*?) ($composed | $prohibited_jis | \z)}osx;
	    } else {
		$regexp = qr{\A (.*?) ($composed | \z)}osx;
	    }
	}

	$Encode::Encoding{$name} = bless {
	    Name => $name,
	    alt => $alt,
	    encoding => $Encode::Encoding{"$encoding-canonic"},
	    jisx0213 => $jisx0213,
	    regexp => $regexp,
	} => __PACKAGE__;
    }
}

# substitution cacharcter for multibyte.
my $subChar = "\x22\x2E"; # GETA MARK

sub encode {
    my ($self, $utf8, $chk) = @_;

    my $chk_sub;
    if (ref $chk eq 'CODE') {
	$chk_sub = $chk;
	$chk = $LEAVE_SRC;
    }
    my $regexp = $self->{regexp};

    my $str = '';

  CHUNKS:
    while ($utf8 =~ /./os) {
	my $errChar = undef;

	while ($utf8 =~ s/$regexp//) {
	    my ($chunk, $mc) = ($1, $2);
	    last CHUNKS unless $chunk =~ /./os or $mc =~ /./os;

	    if ($chunk =~ /./os) {
		$str .= $self->{encoding}->encode($chunk, $FB_QUIET);
		if ($chunk =~ /./os) {
		    $utf8 = $chunk . $mc . $utf8;
		    last;
		}
	    }

	    unless ($mc =~ /./os) {
		next;
	    } elsif ($self->{jisx0213} and $composed{$mc}) {
		$str .= $composed{$mc};
		next;
	    } else {
		$errChar = $mc;
		$utf8 = $mc . $utf8;
		last;
	    }
	}

	$errChar = substr($utf8, 0, 1) unless defined $errChar;

	if ($chk & $DIE_ON_ERR) {
	    croak sprintf $err_encode_nomap, '}\x{', $errChar, $self->{Name};
	}
	if ($chk & $WARN_ON_ERR) {
	    carp sprintf $err_encode_nomap, '}\x{', $errChar, $self->{Name};
	}
	if ($chk & $RETURN_ON_ERR) {
	    last CHUNKS;
	}

	if ($chk_sub) {
	    $str .= join '', map { $chk_sub->(ord $_) } split //, $errChar;
	} else {
	    $str .= $subChar;
	}
	substr($utf8, 0, length $errChar) = '';
    } # CHUNKS

    $_[1] = $utf8 unless $chk & $LEAVE_SRC;
    return $str;
}

sub decode {
    my ($self, $str, $chk) = @_;

    if ($self->{alt} and not ref $chk) {
	$chk &= ~($PERLQQ | $XMLCREF | $HTMLCREF);
    }
    my $utf8 = $self->{encoding}->decode($str, $chk);
    if ($self->{alt} eq 'ascii') {
	$utf8 =~ tr/\x21-\x7E/\x{FF01}-\x{FF5E}/;
    } elsif ($self->{alt} eq 'jis') {
	$utf8 =~ tr/\x21-\x5B\x{00A5}\x5D-\x7D\x{203E}/\x{FF01}-\x{FF3B}\x{FFE5}\x{FF3D}-\x{FF5D}\x{FFE3}/;
    }

    $_[1] = $str unless ref $chk or $chk & $LEAVE_SRC;
    return $utf8;
}

sub perlio_ok { 0 }

1;
__END__

=head1 NAME

Encode::JISX0213::CCS - JIS X 0213 coded character sets

=head1 ABSTRACT

This module provides followng coded character sets.

  reg# Name                    Description
  ----------------------------------------------------------------
   87  jis-x-0208              JIS X 0208-1983, 2nd rev. of JIS X 0208
  168      ditto               JIS X 0208-1990, 3rd rev. of JIS X 0208
       jis-x-0208-ascii
       jis-x-0208-jis
  233  jis-x-0213-plane1       JIS X 0213:2004 level 3 (plane 1)
       jis-x-0213-plane1-ascii
       jis-x-0213-plane1-jis
  228  jis-x-0213-plane1-2000  JIS X 0213:2000 level 3 (plane 1)
       jis-x-0213-plane1-2000-ascii
       jis-x-0213-plane1-2000-jis
  229  jis-x-0213-plane2       JIS X 0213:2000/2004 level 4 (plane 2)

   -   jis-x-0208-0213         Common set of JIS X 0208 and JIS X 0213,
       jis-x-0208-0213-ascii   according to JIS X 0213:2004 Annex 2
       jis-x-0208-0213-jis
  ----------------------------------------------------------------

=head1 DESCRIPTION

To find out how to use this module in detail,
see L<Encode> and L<Encode::ISO2022>.

=head2 Note on Variants

Those suffixed "-ascii" and "-jis" use alternative names for the characters
compatible to ISO/IEC 646 IRV (virtually ASCII) and JIS X 0201 Latin set,
respectively.

=head2 Compatibility

C<jis-x-0208*> include a fallback mapping for HORIZONTAL BAR.
Though it is not normative mapping defined by JIS X 0208,
it is added for compatibility to C<jis0208-raw> encoding in L<Encode::JP>
core module.

However, C<jis-x-0208-0213*> no longer include this mapping.

=head1 SEE ALSO

L<Encode>, L<Encode::ISO2022>, L<Encode::ISO2022::CCS>.

=head1 AUTHOR

Hatuka*nezumi - IKEDA Soji <hatuka(at)nezumi.nu>

=head1 COPYRIGHT

Copyright (C) 2013 Hatuka*nezumi - IKEDA Soji.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut


Makefile.PL
#-*- perl -*-
#-*- coding: us-ascii -*-

use 5.7.2;
use strict;
use warnings;
use Config;
use ExtUtils::MakeMaker;

my %tables = (
    raw_t => [
	qw(jis-x-0208.ucm jis-x-0208-0213.ucm
	jis-x-0213-plane1-2000.ucm jis-x-0213-plane1.ucm
	jis-x-0213-plane2.ucm)
    ],
    sjis_t => [qw(jis-x-0213-annex1.ucm)],
);

unless ($ENV{AGGREGATE_TABLES}) {
    my @ucm;
    for my $k (keys %tables) {
	push @ucm, @{$tables{$k}};
    }
    %tables = ();
    my $seq = 0;
    for my $ucm (sort @ucm) {

	# 8.3 compliance !
	my $t = sprintf("%s_%02d_t", substr($ucm, 0, 2), $seq++);
	$tables{$t} = [$ucm];
    }
}

my ($enc2xs, $encode_h) = ('enc2xs', '.');
my @path_ext = ('');
@path_ext = split(';', $ENV{PATHEXT}) if $^O eq 'MSWin32';
PATHLOOP:
for my $d (@Config{qw/bin sitebin vendorbin/},
        (split /$Config{path_sep}/o, $ENV{PATH})){
    for my $f (qw/enc2xs enc2xs5.7.3/){
        my $path = File::Spec->catfile($d, $f);
        for my $ext (@path_ext) {
            my $bin = "$path$ext";
            -r "$bin" and $enc2xs = $bin and last PATHLOOP;
        }
    }
}
$enc2xs or die "enc2xs not found!";
print "enc2xs is $enc2xs\n";

my %encode_h = ();
for my $d (@INC) {
    my $dir  = File::Spec->catfile($d,   "Encode");
    my $file = File::Spec->catfile($dir, "encode.h");
    -f $file and $encode_h{$dir} = -M $file;
}
%encode_h or die "encode.h not found!";

# find the latest one
($encode_h) = sort { $encode_h{$b} <=> $encode_h{$a} } keys %encode_h;
print "encode.h is at $encode_h\n";

# get XS version.
open my $pm_file, 'lib/Encode/JISX0213/CCS.pm';
my $xs_version = eval [grep {s/^our \$VERSION\s*=\s//} <$pm_file>]->[0];
close $pm_file;

my $name = 'JISX0213';

WriteMakefile(
    INC           => "-I$encode_h",
    NAME          => 'Encode::' . $name,
    VERSION_FROM  => 'lib/Encode/JISX0213.pm',
    ABSTRACT_FROM => 'lib/Encode/JISX0213.pm',
    AUTHOR        => 'Hatuka*nezumi - IKEDA Soji <hatuka@nezumi.nu>',
    LICENSE       => 'perl',
    CONFIGURE_REQUIRES => { 'ExtUtils::MakeMaker' => 0, },
    BUILD_REQUIRES     => {
        'ExtUtils::MakeMaker' => 0,
        'Test::More'          => 0,
    },
    PREREQ_PM     => {
        'Encode::ISO2022'     => '0.03',
    },
    META_MERGE    => {
        resources     => {
            license       => 'http://dev.perl.org/licenses/',
            repository    => 'http://hatuka.nezumi.nu/repos/Encode-JISX0213/',
        },
    },

    XS_VERSION => $xs_version,
    OBJECT  => '$(O_FILES)',
    'dist'  => {
	COMPRESS     => 'gzip -9f',
	SUFFIX       => 'gz',
	DIST_DEFAULT => 'all tardist',
    },
    #MAN3PODS => {},
    # OS 390 winges about line numbers > 64K ???
    XSOPT => '-nolinenumbers',
);

package MY;

sub post_initialize {
    my ($self) = @_;
    my %o;
    my $x = $self->{'OBJ_EXT'};

    # Add the table O_FILES
    foreach my $e (keys %tables) {
	$o{$e . $x} = 1;
    }
    $o{"$name$x"} = 1;
    $self->{'O_FILES'} = [sort keys %o];
    my @files = ("$name.xs");
    $self->{'C'} = ["$name.c"];
    $self->{SOURCE} .= " $name.c"
	if $^O eq 'MacOS' && $self->{SOURCE} !~ /\b$name\.c\b/;
    #$self->{'H'} = [$self->catfile($self->updir,'Encode', 'encode.h')];
    my %xs;
    foreach my $table (keys %tables) {
	push(@{$self->{'C'}}, "$table.c");
	# Do NOT add $table.h etc. to H_FILES unless we own up as to how they
	# get built.
	foreach my $ext (qw($(OBJ_EXT) .c .h .exh .fnm)) {
	    push(@files, $table . $ext);
	}
	$self->{SOURCE} .= " $table.c"
	    if $^O eq 'MacOS' && $self->{SOURCE} !~ /\b$table\.c\b/;
    }
    $self->{'XS'} = {"$name.xs" => "$name.c"};
    $self->{'clean'}{'FILES'} .= join(' ', @files);
    open(XS, ">$name.xs") || die "Cannot open $name.xs:$!";
    print XS <<'END';
#include <EXTERN.h>
#include <perl.h>
#include <XSUB.h>
#define U8 U8
#include "encode.h"
END
    foreach my $table (sort keys %tables) {
	print XS qq[#include "${table}.h"\n];
    }
    print XS <<"END";

static void
Encode_XSEncoding(pTHX_ encode_t *enc)
{
 dSP;
 HV *stash = gv_stashpv("Encode::XS", TRUE);
 SV *sv    = sv_bless(newRV_noinc(newSViv(PTR2IV(enc))),stash);
 int i = 0;
 PUSHMARK(sp);
 XPUSHs(sv);
 while (enc->name[i])
  {
   const char *name = enc->name[i++];
   XPUSHs(sv_2mortal(newSVpvn(name,strlen(name))));
  }
 PUTBACK;
 call_pv("Encode::define_encoding",G_DISCARD);
 SvREFCNT_dec(sv);
}

MODULE = Encode::$name	PACKAGE = Encode::$name
PROTOTYPES: DISABLE
BOOT:
{
END
    foreach my $table (sort keys %tables) {
	print XS qq[#include "${table}.exh"\n];
    }
    print XS "}\n";
    close(XS);
    return "# Built $name.xs\n\n";
}

sub postamble {
    my $self = shift;
    my $dir = 'ucm';
    my $str =
	"# $name\$(OBJ_EXT) depends on .h and .exh files not .c files - but all written by enc2xs\n";
    $str .= "$name.c : $name.xs ";
    foreach my $table (keys %tables) {
	$str .= " $table.c";
    }
    $str .= "\n\n";
    $str .= "$name\$(OBJ_EXT) : $name.c\n\n";

    foreach my $table (keys %tables) {
	my $numlines    = 1;
	my $lengthsofar = length($str);
	my $continuator = '';
	$str .= "$table.c : $enc2xs Makefile.PL";
	foreach my $file (@{$tables{$table}}) {
	    $str .= $continuator . ' ' . $self->catfile($dir, $file);
	    if (length($str) - $lengthsofar > 128 * $numlines) {
		$continuator .= " \\\n\t";
		$numlines++;
	    } else {
		$continuator = '';
	    }
	}
	my $plib = $self->{PERL_CORE} ? '"-I$(PERL_LIB)"' : '';
	$plib .= " -MCross=$::Cross::platform" if defined $::Cross::platform;
	my $ucopts = '-"Q"';
	$str .=
	    qq{\n\t\$(PERL) $plib $enc2xs $ucopts -o \$\@ -f $table.fnm\n\n};
	open(FILELIST, ">$table.fnm") ||
	    die "Could not open $table.fnm: $!";
	foreach my $file (@{$tables{$table}}) {
	    print FILELIST $self->catfile($dir, $file) . "\n";
	}
	close(FILELIST);
    }
    return $str;
}

